lib/wallaby/browser.ex

```diff
@@ -0,0 +1,831 @@
+  @spec take_screenshot(parent) :: parent
+
+  def take_screenshot(screenshotable) do
+    image_data =
+      screenshotable
+      |> Driver.take_screenshot
+
+    path = path_for_screenshot
+    File.write! path, image_data
+
+    Map.update(screenshotable, :screenshots, [], &(&1 ++ [path]))
+  end
+
+  @doc """
+  Gets the size of the session's window.
+  """
+  @spec window_size(parent) :: %{String.t => pos_integer, String.t => pos_integer}
+
+  def window_size(session) do
+    {:ok, size} = Driver.get_window_size(session)
+    size
+  end
+
+  @doc """
+  Sets the size of the sessions window.
+  """
+  @spec resize_window(parent, pos_integer, pos_integer) :: parent
+
+  def resize_window(session, width, height) do
+    {:ok, _} = Driver.set_window_size(session, width, height)
+    session
+  end
+
+  @doc """
+  Gets the current url of the session
+  """
+  @spec current_url(parent) :: String.t
+
+  def current_url(parent) do
+    Driver.current_url!(parent)
+  end
+
+  @doc """
+  Gets the current path of the session
+  """
+  @spec current_path(parent) :: String.t
+
+  def current_path(parent) do
+    Driver.current_path!(parent)
+  end
+
+  @doc """
+  Gets the title for the current page
+  """
+  @spec page_title(parent) :: String.t
+
+  def page_title(session) do
+    {:ok, title} = Driver.page_title(session)
+    title
+  end
+
+  @doc """
+  Executes javascript synchoronously, taking as arguments the script to execute,
+  and optionally a list of arguments available in the script via `arguments`
+  """
+  @spec execute_script(parent, String.t, list) :: parent
+
+  def execute_script(session, script, arguments \\ []) do
+    {:ok, value} = Driver.execute_script(session, script, arguments)
+    value
+  end
+
+  @doc """
+  Sends a list of key strokes to active element. If strings are included
+  then they are sent as individual keys. Special keys should be provided as a
+  list of atoms, which are automatically converted into the corresponding key
+  codes.
+
+  For a list of available key codes see `Wallaby.Helpers.KeyCodes`.
+
+  ## Example
+
+      iex> Wallaby.Session.send_keys(session, ["Example Text", :enter])
+      iex> Wallaby.Session.send_keys(session, [:enter])
+      iex> Wallaby.Session.send_keys(session, [:shift, :enter])
+  """
+  @spec send_keys(parent, list(atom)) :: parent
+  @spec send_keys(parent, StatelessQuery.t, list(atom)) :: parent
+
+  def send_keys(parent, query, list) do
+    parent
+    |> find(query)
+    |> click()
+    |> send_keys(list)
+  end
+  def send_keys(parent, keys) when is_list(keys) do
+    {:ok, _} = Driver.send_keys(parent, keys)
+    parent
+  end
+
+  @doc """
+  Retrieves the source of the current page.
+  """
+  @spec page_source(parent) :: String.t
+
+  def page_source(session) do
+    {:ok, source} = Driver.page_source(session)
+    source
+  end
+
+  @doc """
+  Sets the value of an element.
+  """
+  def set_value(element, value) do
+    {:ok, _} = Driver.set_value(element, value)
```

Comment:
I think it makes sense to return the element here so actions can be chained. All of the explicit pattern matches to `{:ok, _}` need to be turned into case statements so that the browser handles errors correctly. Otherwise people could get internal wallaby errors thrown in their tests.